### ISSUE TITLE
ASan does not like Variant

### ISSUE THREAD

**Issue by ivan-cukic** [firebase/firebase-cpp-sdk#16]
Address sanitizer (android x86, NDK19) reports an error in Variant's move constructor:

```
07-17 07:37:03.890 31772 31772 I fbuserapp: 
07-17 07:37:03.890 31772 31772 I fbuserapp: 0xa0f02b40 is located 0 bytes inside of 60-byte region [0xa0f02b40,0xa0f02b7c)
07-17 07:37:03.890 31772 31772 I fbuserapp: 
07-17 07:37:03.890 31772 31772 I fbuserapp: ==31772==AddressSanitizer CHECK failed: /usr/local/google/buildbot/src/android/llvm-toolchain/toolchain/compiler-rt/lib/asan/asan_descriptions.cc:176 "((id)) != (0)" (0x0, 0x0)
```

Any ideas of how to handle this?

This is the relevant backtrace:

```
********** Crash dump: **********
Build fingerprint: 'google/sdk_gphone_x86/generic_x86:9/PSR1.180720.093/5456446:userdebug/dev-keys'
#00 0x00000b39 [vdso:ec404000] (__kernel_vsyscall+9)                                                                                                                                                                                                                                                          
#01 0x0001fdf8 /system/lib/libc.so (syscall+40)                                                                                                                                                                                                                                                               
#02 0x00022ed3 /system/lib/libc.so (abort+115)                                                                                                         
#03 0x0005d451 /data/app/fbuserapp-tW8SRDlmxN88zZO-DIwvmQ==/lib/x86/libclang_rt.asan-i686-android.so (offset 0x52000)                                                                                                                                                                                      
                                                                                                         ??                                                                                                                                                                                                   
                                                                                                         ??:0:0
#04 0x0005c0bd /data/app/fbuserapp-tW8SRDlmxN88zZO-DIwvmQ==/lib/x86/libclang_rt.asan-i686-android.so (offset 0x52000)                     
                                                                                                         ??                              
                                                                                                         ??:0:0
#05 0x000dbb09 /data/app/fbuserapp-tW8SRDlmxN88zZO-DIwvmQ==/lib/x86/libclang_rt.asan-i686-android.so (offset 0x52000)                                                                                                                                                                                      
                                                                                                         ??                                                                                                                                                                                                   
                                                                                                         ??:0:0
#06 0x0005c143 /data/app/fbuserapp-tW8SRDlmxN88zZO-DIwvmQ==/lib/x86/libclang_rt.asan-i686-android.so (offset 0x52000)
                                                                                                         ??
                                                                                                         ??:0:0
#07 0x00071152 /data/app/fbuserapp-tW8SRDlmxN88zZO-DIwvmQ==/lib/x86/libclang_rt.asan-i686-android.so (offset 0x52000)
                                                                                                         ??
                                                                                                         ??:0:0
#08 0x0007269d /data/app/fbuserapp-tW8SRDlmxN88zZO-DIwvmQ==/lib/x86/libclang_rt.asan-i686-android.so (offset 0x52000)
                                                                                                         ??
                                                                                                         ??:0:0
#09 0x000743c6 /data/app/fbuserapp-tW8SRDlmxN88zZO-DIwvmQ==/lib/x86/libclang_rt.asan-i686-android.so (offset 0x52000)                                                                                                                                                                                      
                                                                                                         ??                                                                                                                                                                                                   
                                                                                                         ??:0:0
#10 0x000d98e4 /data/app/fbuserapp-tW8SRDlmxN88zZO-DIwvmQ==/lib/x86/libclang_rt.asan-i686-android.so (offset 0x52000)
                                                                                                         ??
                                                                                                         ??:0:0
#11 0x000d72e5 /data/app/fbuserapp-tW8SRDlmxN88zZO-DIwvmQ==/lib/x86/libclang_rt.asan-i686-android.so (offset 0x52000)
                                                                                                         ??
                                                                                                         ??:0:0
#12 0x000d929a /data/app/fbuserapp-tW8SRDlmxN88zZO-DIwvmQ==/lib/x86/libclang_rt.asan-i686-android.so (offset 0x52000)                                                                                                                                                                                      
                                                                                                         ??                                                                                                                                                                                                   
                                                                                                         ??:0:0
#13 0x000d9f38 /data/app/fbuserapp-tW8SRDlmxN88zZO-DIwvmQ==/lib/x86/libclang_rt.asan-i686-android.so (offset 0x52000) (__asan_report_store4+40)
                                                                                                         __asan_report_store4
                                                                                                         ??:0:0
#14 0x00338e66 /data/app/fbuserapp-tW8SRDlmxN88zZO-DIwvmQ==/lib/x86/libFBUserApp.so (firebase::Variant::Variant(firebase::Variant&&)+118)
                                                                                          firebase::Variant::Variant(firebase::Variant&&)
                                                                                          ??:0:0
#15 0x007f182c /data/app/fbuserapp-tW8SRDlmxN88zZO-DIwvmQ==/lib/x86/libFBUserApp.so                                                                                                                                                                                                                      
??                                                                                                                                                                                                                                                                                                            
??:0:0                                                                                                                                                 
#16 0x007f0dad /data/app/fbuserapp-tW8SRDlmxN88zZO-DIwvmQ==/lib/x86/libFBUserApp.so (firebase::database::internal::JavaObjectToVariant(_JNIEnv*, _jobject*)+1037)                                                                 
                                                                                          firebase::database::internal::JavaObjectToVariant(_JNIEnv*, _jobject*)                                                                  
                                                                                          ??:0:0
#17 0x007f1c42 /data/app/fbuserapp-tW8SRDlmxN88zZO-DIwvmQ==/lib/x86/libFBUserApp.so
??                     
??:0:0
#18 0x007f0ec4 /data/app/fbuserapp-tW8SRDlmxN88zZO-DIwvmQ==/lib/x86/libFBUserApp.so (firebase::database::internal::JavaObjectToVariant(_JNIEnv*, _jobject*)+1316)
                                                                                          firebase::database::internal::JavaObjectToVariant(_JNIEnv*, _jobject*)
                                                                                          ??:0:0
```

**Comment by chkuang-g** [comment]
@ivan-cukic 

Thank you for reporting this!

Could you provide a minimal sample that failed with ASan?  Also, could you let use know what kind of dataase value that is causing this issue?

Thank you!
Shawn

**Comment by a-maurice** [comment]
Hi @ivan-cukic 

This issue should now be closed with the release of 6.4.0, available https://github.com/firebase/firebase-cpp-sdk/releases/tag/v6.4.0

Thanks for bringing it to our attention!

**Comment by ivan-cukic** [comment]
Thanks!

Sorry for not posting the minimal samle - the comment somehow got through unnoticed.
